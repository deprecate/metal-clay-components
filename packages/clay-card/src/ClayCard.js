import './ClayCardBase';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayCard.soy.js';

/**
 * Metal ClayCard component.
 */
class ClayCard extends Component {}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayCard.STATE = {
	/**
	 * Actions in the dropdown on card.
	 * @instance
	 * @memberof ClayCard
	 * @type {?array|undefined}
	 * @default undefined
	 */
	actionsItems: Config.arrayOf(
		Config.shapeOf({
			label: Config.string().required(),
			separator: Config.bool().value(false),
			url: Config.string(),
		})
	),

	/**
	 * Transform card with checkbox.
	 * @instance
	 * @memberof ClayCard
	 * @type {?array|undefined}
	 * @default undefined
	 */
	checkbox: Config.shapeOf({
		checked: Config.bool().value(false),
		disabled: Config.bool().value(false),
		indeterminate: Config.bool().value(false),
		name: Config.string(),
		value: Config.string(),
	}),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * Render ClayIcon in the background element.
	 * @instance
	 * @memberof ClayCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	icon: Config.string(),

	/**
	 * Alt the image.
	 * @instance
	 * @memberof ClayCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	imageAlt: Config.string(),

	/**
	 * The src to be applied to the background or user card image.
	 * @instance
	 * @memberof ClayCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	imageSrc: Config.string(),

	/**
	 * Render label in the ClayLabel element.
	 * @instance
	 * @memberof ClayCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	label: Config.string(),

	/**
	 * Style of the ClayLabel.
	 * @instance
	 * @memberof ClayCard
	 * @type {?string}
	 * @default secondary
	 */
	labelStyle: Config.oneOf([
		'danger',
		'info',
		'secondary',
		'success',
		'warning',
	]).value('secondary'),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayCard
	 * @type {!string}
	 * @default undefined
	 */
	spritemap: Config.string().required(),

	/**
	 * Subtitle of the card.
	 * @instance
	 * @memberof ClayCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	subtitle: Config.string(),

	/**
	 * Title of the card.
	 * @instance
	 * @memberof ClayCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	title: Config.string(),

	/**
	 * Render label in the ClaySticker element.
	 * @instance
	 * @memberof ClayCard
	 * @type {?string|undefined}
	 * @default undefined
	 */
	type: Config.string(),

	/**
	 * Style of the ClaySticker element.
	 * @instance
	 * @memberof ClayCard
	 * @type {?string}
	 * @default primary
	 */
	typeStyle: Config.oneOf([
		'danger',
		'dark',
		'info',
		'light',
		'primary',
		'secondary',
		'success',
		'warning',
	]).value('primary'),

	/**
	 * Transform title in the url.
	 * @instance
	 * @memberof ClayCard
	 * @type {?string}
	 * @default undefined
	 */
	url: Config.string(),
};

defineWebComponent('clay-card', ClayCard);

Soy.register(ClayCard, templates);

export {ClayCard};
export default ClayCard;
