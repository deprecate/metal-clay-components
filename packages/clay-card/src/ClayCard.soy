{namespace ClayCard}

{template .render}
    {@param? actionItems: list<[
        label: string,
        separator: bool,
        url: string
    ]>}
    {@param? disabled: bool}
    {@param? elementClasses: string}
    {@param? icon: string}
    {@param? id: string}
    {@param? imageAlt: string}
	{@param? imageSrc: string}
    {@param? inputName: string}
    {@param? inputValue: string}
    {@param? label: string}
	{@param? labelStyle: string}
    {@param? selectable: bool}
    {@param? selected: bool}
    {@param? spritemap: string}
    {@param? subtitle: string}
    {@param? title: string}
    {@param? type: string}
	{@param? typeStyle: string}
    {@param? url: string}


    {call ClayCardBase.render}
        {param actionItems: $actionItems /}
        {param disabled: $disabled /}
        {param elementClasses: $elementClasses /}
        {param icon: $icon /}
        {param id: $id /}
        {param imageAlt: $imageAlt /}
        {param imageSrc: $imageSrc /}
        {param inputName: $inputName /}
        {param inputValue: $inputValue /}
        {param label: $label /}
        {param labelStyle: $labelStyle /}
        {param ref: 'card' /}
        {param selectable: $selectable /}
        {param selected: $selected /}
        {param spritemap: $spritemap /}
        {param subtitle: $subtitle /}
        {param title: $title /}
        {param type: $type /}
        {param typeStyle: $typeStyle /}
        {param url: $url /}
        {param variant: 'image' /}
    {/call}
{/template}

/**
 * Render the visual area of the card.
 */
{deltemplate ClayCardBase.VisualArea variant="'image'"}
	{@param? icon: string}
	{@param? imageAlt: string}
	{@param? imageSrc: string}
    {@param? spritemap: string}
	{@param? type: string}
	{@param? typeStyle: string}

	<div class="aspect-ratio card-item-first bg-checkered">
		{let $aspectRatioClasses kind="text"}
			aspect-ratio-item-center-middle aspect-ratio-item-fluid
		{/let}

		{if $imageSrc}
			{let $imageAttributes kind="attributes"}
				alt="
					{if $imageAlt}
						{$imageAlt}
					{/if}
				"

				class="{$aspectRatioClasses}"
				src="{$imageSrc}"
			{/let}

			<img {$imageAttributes} />
		{elseif $icon and $spritemap}
			<div class="{$aspectRatioClasses}">
				{call ClayIcon.render}
					{param spritemap: $spritemap /}
					{param symbol: $icon /}
				{/call}
			</div>
		{/if}

		{if $type}
			{call ClaySticker.render}
				{param label: $type /}
				{param position: 'bottom-left' /}
				{param style: $typeStyle ?: 'primary' /}
			{/call}
		{/if}
	</div>
{/deltemplate}