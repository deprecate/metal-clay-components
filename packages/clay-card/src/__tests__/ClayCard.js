import ClayCard from '../ClayCard';

let component;
let spritemap = 'icons.svg';
let actionItems = [
	{
		label: 'Edit',
		url: '#1',
	},
	{
		label: 'Save',
		url: '#1',
	},
];

describe('ClayCard', function() {
	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('should render the default markup', () => {
		component = new ClayCard({
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with classes', () => {
		component = new ClayCard({
			spritemap: spritemap,
			elementClasses: 'my-custom-class',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with id', () => {
		component = new ClayCard({
			spritemap: spritemap,
			id: 'myId',
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with actionItems', () => {
		component = new ClayCard({
			actionItems: actionItems,
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a selected ClayCard', () => {
		component = new ClayCard({
			selectable: true,
			selected: true,
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a disabled ClayCard', () => {
		component = new ClayCard({
			disabled: true,
			selectable: true,
			selected: true,
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with input `name`', () => {
		component = new ClayCard({
			inputName: 'checkbox01',
			selectable: true,
			selected: true,
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with input `value`', () => {
		component = new ClayCard({
			inputName: 'checkbox',
			selectable: true,
			selected: true,
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with icon', () => {
		component = new ClayCard({
			icon: 'documents',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with imageAlt', () => {
		component = new ClayCard({
			imageAlt: 'thumbnail',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with imageSrc', () => {
		component = new ClayCard({
			imageSrc: 'thumbnail_coffee.jpg',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with label', () => {
		component = new ClayCard({
			label: 'Approved',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with labelStyle `danger color`', () => {
		component = new ClayCard({
			label: 'Approved',
			labelStyle: 'danger',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with labelStyle `info color`', () => {
		component = new ClayCard({
			label: 'Approved',
			labelStyle: 'info',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with labelStyle `secondary color`', () => {
		component = new ClayCard({
			label: 'Approved',
			labelStyle: 'secondary',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with labelStyle `success color`', () => {
		component = new ClayCard({
			label: 'Approved',
			labelStyle: 'success',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with labelStyle `warning color`', () => {
		component = new ClayCard({
			label: 'Approved',
			labelStyle: 'warning',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with subtitle', () => {
		component = new ClayCard({
			subtitle: 'Author Action',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with title', () => {
		component = new ClayCard({
			title: 'empty-background.png',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with type', () => {
		component = new ClayCard({
			type: 'DOC',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with type `danger color`', () => {
		component = new ClayCard({
			type: 'DOC',
			typeStyle: 'danger',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with type `dark color`', () => {
		component = new ClayCard({
			type: 'DOC',
			typeStyle: 'dark',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with type `info color`', () => {
		component = new ClayCard({
			type: 'DOC',
			typeStyle: 'info',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with type `light color`', () => {
		component = new ClayCard({
			type: 'DOC',
			typeStyle: 'light',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with type `primary color`', () => {
		component = new ClayCard({
			type: 'DOC',
			typeStyle: 'primary',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with type `secondary color`', () => {
		component = new ClayCard({
			type: 'DOC',
			typeStyle: 'secondary',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with type `success color`', () => {
		component = new ClayCard({
			type: 'DOC',
			typeStyle: 'success',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with type `warning color`', () => {
		component = new ClayCard({
			type: 'DOC',
			typeStyle: 'warning',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});

	it('should render a ClayCard with url', () => {
		component = new ClayCard({
			url: '#1',
			spritemap: spritemap,
		});

		expect(component).toMatchSnapshot();
	});
});
