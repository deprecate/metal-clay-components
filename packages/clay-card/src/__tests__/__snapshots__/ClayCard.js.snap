// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayCard should render a ClayCard with classes 1`] = `
<div class="card card-type-asset my-custom-class">
  <div class="aspect-ratio card-item-first bg-checkered">
    <span class="sticker sticker-danger sticker-bottom-left">PNG</span>
  </div>
  <div class="card-body">
    <div class="card-row">
      <div class="flex-col flex-col-expand">
        <div class="card-title text-truncate"></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayCard should render a ClayCard with id 1`] = `
<div class="card card-type-asset" id="myId">
  <div class="aspect-ratio card-item-first bg-checkered">
    <span class="sticker sticker-danger sticker-bottom-left">PNG</span>
  </div>
  <div class="card-body">
    <div class="card-row">
      <div class="flex-col flex-col-expand">
        <div class="card-title text-truncate"></div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ClayCard should render the default markup 1`] = `
<div class="card card-type-asset">
  <div class="aspect-ratio card-item-first bg-checkered">
    <span class="sticker sticker-danger sticker-bottom-left">PNG</span>
  </div>
  <div class="card-body">
    <div class="card-row">
      <div class="flex-col flex-col-expand">
        <div class="card-title text-truncate"></div>
      </div>
    </div>
  </div>
</div>
`;
